### PR TITLE
Refactor the code

### PR DESCRIPTION
**Pull request by pawel-czyz**
This PR refactors the code in the following ways:

  - [ ] Resolve issue #24, using consistent shape `(n_timepoints, n_variants)`
  - [ ] Add more type annotations.
  - [ ] Refactor the notebook, using more informative variable names and auxiliary utilities to simplify it.

### CODE REVIEW

File: src/covvfit/plotting/_grid.py [hunk 1]
```diff
@@ -58,3 +60,119 @@ def plot_grid(
                 over_diag(ax, i, j)
 
     return fig, axes
+
+
+@dataclass(frozen=False)
```
Comment (pawel-czyz) [inline src/covvfit/plotting/_grid.py:65]: In the notebook we are fortunate to use eight cities arranged in two columns.

I wanted to create a utility which automatically arranges a list of plots (one per city) in a grid and has convenient utilities to iterate over them.

For example, if we had 9 cities, they would be arranged in a 5 x 2 grid with 9 active axes and one inactive (and invisible) axis.

File: src/covvfit/plotting/_grid.py [hunk 2]
```diff
@@ -58,3 +60,119 @@ def plot_grid(
...
+            `subplots_from_axsize`. For example,
+            ```
+            wspace=0.2,  # Changes the horizontal spacing
+            hspace=0.3,  # Changes the vertical spacing
+            left=0.5,    # Changes the left margin
+            ```
+    """
+    nrows = _calculate_nrows(nplots, ncols=ncols)
+
+    fig, axs = subplots_from_axsize(
```
Comment (pawel-czyz) [inline src/covvfit/plotting/_grid.py:162]: This function comes from a [handy small package](https://pypi.org/project/subplots-from-axsize/) created by Frederic. 

I added it as a dependency to `pyproject.toml`, so you may need to install it with Poetry (`poetry update`) or PIP (`pip install subplots-from-axsize`).

File: src/covvfit/plotting/_timeseries.py [hunk 3]
```diff
@@ -2,10 +2,15 @@
 
 import matplotlib.lines as mlines
 import matplotlib.patches as mpatches
+import matplotlib.pyplot as plt
 import numpy as np
 import pandas as pd
+from jaxtyping import Array, Float
 
-colors_covsp = {
+Variant = str
```
Comment (pawel-czyz) [inline src/covvfit/plotting/_timeseries.py:10]: This annotations is just for total beginner virologist like myself, so I don't confuse variants with color names.

File: examples/frequentist_notebook_jax.py [hunk 4]
```diff
@@ -90,156 +92,185 @@ def match_date(start_date):
...
-    #     # plot 1-fitted and predicted values
-    plot_complement(ax, ts_lst[i], y_fit_lst[i], variants)
-    #     plot_complement(ax, ts_pred_lst[i], y_pred_lst[i], variants, linetype="--")
-    # plot raw deconvolved values
-    plot_data(ax, ts_lst[i], ys_lst2[i], variants, colors)
-    # make confidence bands and plot them
-    conf_bands = y_fit_lst_confint[i]
-    plot_confidence_bands(
+def plot_city(ax, i: int) -> None:
+    def remove_0th(arr):
```
Comment (pawel-czyz) [inline examples/frequentist_notebook_jax.py:244]: Note that we don't define any auxiliary arrays for the "variants under investigation"/"real variants" anymore: we always work with the "effective variants", i.e., with an additional auxiliary "other" variant.

Only at the end, when  we want to plot only the "real variants", we remove it using this slicing function.
Doing it in this way, we have fewer arrays to remember about.

File: src/covvfit/_preprocess_abundances.py [hunk 5]
```diff
@@ -13,34 +14,56 @@ def preprocess_df(
     df: pd.DataFrame,
     cities: list[str],
     variants: list[str],
-    undertermined_thresh: float = 0.01,
+    *,
```
Comment (pawel-czyz) [inline src/covvfit/_preprocess_abundances.py:17]: I added "*", [making the rest of the arguments keyword-only](https://peps.python.org/pep-0570/#keyword-only-arguments), which hopefully will limit the number of bugs introduced by accidentally swapping `date_min` and `date_max` or column names.

File: src/covvfit/plotting/_grid.py (hunk 1 above)
Comment (dr-david) [reply src/covvfit/plotting/_grid.py:65]: super useful !

File: src/covvfit/plotting/_grid.py (hunk 2 above)
Comment (dr-david) [reply src/covvfit/plotting/_grid.py:162]: Thanks Frederic :) super useful
